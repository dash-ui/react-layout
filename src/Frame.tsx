import * as React from 'react'
import css from 'minify-css.macro'
import clsx from 'clsx'
import {useLayout} from './Layout'
import type {DefaultVars} from '@-ui/react'
import type {MediaQueryProp} from './Layout'

export const Frame = React.forwardRef<any, FrameProps>(
  (
    {
      as: As = 'div',
      className,
      display,
      position,
      width,
      height,
      size,
      pad,
      bg,
      elevation,
      radius,
      ...props
    },
    ref
  ) => {
    const {mq} = useLayout()

    return (
      <As
        ref={ref}
        className={clsx(
          className,
          mq.prop(frameStyle, display),
          mq.prop(frameStyle, position),
          mq.prop(widthStyle, width),
          mq.prop(heightStyle, height),
          mq.prop(sizeStyle, size),
          mq.prop(padStyle, pad),
          mq.prop(bgStyle, bg),
          mq.prop(elevationStyle, elevation),
          mq.prop(radiusStyle, radius)
        )}
        {...props}
      />
    )
  }
)

const frameStyle = {
  flex: css`
    display: flex;
  `,
  inlineFlex: css`
    display: inline-flex;
  `,
  block: css`
    display: block;
  `,
  inlineBlock: css`
    display: inline-block;
  `,
  inline: css`
    display: inline;
  `,
  none: css`
    display: none;
  `,
  relative: css`
    position: relative;
  `,
  absolute: css`
    position: absolute;
  `,
  sticky: css`
    position: sticky;
  `,
  fixed: css`
    position: fixed;
  `,
}

const widthStyle = (width: number | string) => ({width})
const heightStyle = (height: number | string) => ({height})
const sizeStyle = (size: number | string) => ({
  width: size,
  height: size,
})

const padStyle = (
  // @ts-ignore
  padProp: keyof DefaultVars['pad'] | (keyof DefaultVars['pad'])[]
) => ({pad}) => {
  if (Array.isArray(padProp)) {
    return {padding: padProp.map((k) => pad[k]).join(' ')}
  }

  return css`
    padding: ${pad[padProp]};
  `
}

// @ts-ignore
const bgStyle = (bg: keyof DefaultVars['color']) => ({color}) => css`
  background: ${color[bg]};
`

// @ts-ignore
const elevationStyle = (elevationProp: keyof DefaultVars['elevation']) => ({
  elevation,
}) => css`
  box-shadow: ${elevation[elevationProp]};
`
// @ts-ignore
const radiusStyle = (radiusProp: keyof DefaultVars['radius']) => ({radius}) => {
  if (Array.isArray(radiusProp)) {
    return {borderRadius: radiusProp.map((k) => radius[k]).join(' ')}
  }

  return css`
    border-radius: ${radius[radiusProp]};
  `
}

export interface FrameProps {
  readonly as?: keyof JSX.IntrinsicElements | React.ComponentType<any>
  readonly className?: string | string[]
  readonly display?: MediaQueryProp<
    'flex' | 'inlineFlex' | 'block' | 'inlineBlock' | 'inline'
  >
  readonly position?: MediaQueryProp<
    'relative' | 'absolute' | 'sticky' | 'fixed'
  >
  readonly width?: MediaQueryProp<number | string>
  readonly height?: MediaQueryProp<number | string>
  readonly size?: MediaQueryProp<number | string>
  readonly pad?: MediaQueryProp<
    // @ts-ignore
    keyof DefaultVars['pad'] | (keyof DefaultVars['pad'])[]
  >
  // @ts-ignore
  readonly bg?: MediaQueryProp<keyof DefaultVars['color']>
  // @ts-ignore
  readonly elevation?: MediaQueryProp<keyof DefaultVars['elevation']>
  readonly radius?: MediaQueryProp<
    // @ts-ignore
    keyof DefaultVars['radius'] | (keyof DefaultVars['radius'])[]
  >
}
