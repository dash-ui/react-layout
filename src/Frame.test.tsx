/* jest */
import * as React from 'react'
import {render} from '@testing-library/react'
import {Frame} from './Frame'
import {silenceErrors, renderMq, mediaQueries} from 'test-utils'

describe('<Frame> without media queries', () => {
  it(
    'should throw if media queries are used without <LayoutProvider>',
    silenceErrors(() => {
      expect(() =>
        render(<Frame display={{phone: 'block'}} />)
      ).toThrowErrorMatchingSnapshot()
    })
  )

  it('applies the "display" prop', () => {
    const {getByTestId} = render(<Frame display="block" data-testid="el" />)
    expect(getByTestId('el')).toHaveStyleRule('display', 'block')
  })

  it('applies the "position" prop', () => {
    const {getByTestId} = render(<Frame position="relative" data-testid="el" />)
    expect(getByTestId('el')).toHaveStyleRule('position', 'relative')
  })

  it('applies the "width" prop', () => {
    const {getByTestId} = render(<Frame width="100%" data-testid="el" />)
    expect(getByTestId('el')).toHaveStyleRule('width', '100%')
  })

  it('applies the "width" prop w/ px default', () => {
    const {getByTestId} = render(<Frame width={100} data-testid="el" />)
    expect(getByTestId('el')).toHaveStyleRule('width', '100px')
  })

  it('applies the "height" prop', () => {
    const {getByTestId} = render(<Frame height="100%" data-testid="el" />)
    expect(getByTestId('el')).toHaveStyleRule('height', '100%')
  })

  it('applies the "height" prop w/ px default', () => {
    const {getByTestId} = render(<Frame height={100} data-testid="el" />)
    expect(getByTestId('el')).toHaveStyleRule('height', '100px')
  })

  it('applies the "width" and "height" props together', () => {
    const {getByTestId} = render(
      <Frame width={200} height={100} data-testid="el" />
    )
    expect(getByTestId('el')).toHaveStyleRule('width', '200px')
    expect(getByTestId('el')).toHaveStyleRule('height', '100px')
  })

  it('applies the "size" prop', () => {
    const {getByTestId} = render(<Frame size="100%" data-testid="el" />)
    expect(getByTestId('el')).toHaveStyleRule('width', '100%')
    expect(getByTestId('el')).toHaveStyleRule('height', '100%')
  })

  it('applies the "size" prop w/ px default', () => {
    const {getByTestId} = render(<Frame size={100} data-testid="el" />)
    expect(getByTestId('el')).toHaveStyleRule('width', '100px')
    expect(getByTestId('el')).toHaveStyleRule('height', '100px')
  })

  it('grants the "size" prop precedence over "width" and "height"', () => {
    const {getByTestId} = render(
      <Frame width="100%" height="100%" size="50%" data-testid="el" />
    )
    expect(getByTestId('el')).toHaveStyleRule('width', '50%')
    expect(getByTestId('el')).toHaveStyleRule('height', '50%')
  })

  it('applies the "bg" prop', () => {
    const {getByTestId} = render(<Frame bg="green" data-testid="el" />)
    expect(getByTestId('el')).toHaveStyleRule(
      'background',
      'var(--color-green)'
    )
  })

  it('applies the "pad" prop', () => {
    const {getByTestId} = render(<Frame pad={1} data-testid="el" />)
    expect(getByTestId('el')).toHaveStyleRule('padding', 'var(--pad-1)')
  })

  it('applies the "pad" prop w/ array', () => {
    const {getByTestId} = render(<Frame pad={[0, 'auto']} data-testid="el" />)
    expect(getByTestId('el')).toHaveStyleRule(
      'padding',
      'var(--pad-0) var(--pad-auto)'
    )
  })

  it('applies the "radius" prop', () => {
    const {getByTestId} = render(<Frame radius="sm" data-testid="el" />)
    expect(getByTestId('el')).toHaveStyleRule(
      'border-radius',
      'var(--radius-sm)'
    )
  })

  it('applies the "radius" prop w/ array', () => {
    const {getByTestId} = render(
      <Frame radius={['sm', 'md']} data-testid="el" />
    )
    expect(getByTestId('el')).toHaveStyleRule(
      'border-radius',
      'var(--radius-sm) var(--radius-md)'
    )
  })

  it('applies the "elevation" prop', () => {
    const {getByTestId} = render(<Frame elevation="low" data-testid="el" />)
    expect(getByTestId('el')).toHaveStyleRule(
      'box-shadow',
      'var(--elevation-low)'
    )
  })

  it('applies several props', () => {
    const {asFragment, getByTestId} = render(
      <Frame
        display="block"
        position="absolute"
        size="200px"
        pad={3}
        elevation="high"
        bg="blue"
        data-testid="el"
      />
    )
    expect(asFragment()).toMatchSnapshot()
    expect(getByTestId('el')).toHaveStyleRule('display', 'block')
    expect(getByTestId('el')).toHaveStyleRule('position', 'absolute')
    expect(getByTestId('el')).toHaveStyleRule('width', '200px')
    expect(getByTestId('el')).toHaveStyleRule('height', '200px')
    expect(getByTestId('el')).toHaveStyleRule('padding', 'var(--pad-3)')
    expect(getByTestId('el')).toHaveStyleRule(
      'box-shadow',
      'var(--elevation-high)'
    )
    expect(getByTestId('el')).toHaveStyleRule('background', 'var(--color-blue)')
  })

  it('works inside the <LayoutProvider>', () => {
    const {getByTestId} = renderMq(<Frame display="block" data-testid="el" />)
    expect(getByTestId('el')).toHaveStyleRule('display', 'block')
  })
})

describe('<Frame> with media queries', () => {
  it('applies the "display" prop', () => {
    const {getByTestId} = renderMq(
      <Frame display={{phone: 'block'}} data-testid="el" />
    )
    expect(getByTestId('el')).toHaveStyleRule('display', 'block', {
      media: mediaQueries.phone,
    })
  })

  it('applies the "position" prop', () => {
    const {getByTestId} = renderMq(
      <Frame position={{phone: 'relative'}} data-testid="el" />
    )
    expect(getByTestId('el')).toHaveStyleRule('position', 'relative', {
      media: mediaQueries.phone,
    })
  })

  it('applies the "width" prop', () => {
    const {getByTestId} = renderMq(
      <Frame width={{phone: '100%'}} data-testid="el" />
    )
    expect(getByTestId('el')).toHaveStyleRule('width', '100%', {
      media: mediaQueries.phone,
    })
  })

  it('applies the "height" prop', () => {
    const {getByTestId} = renderMq(
      <Frame height={{phone: '100%'}} data-testid="el" />
    )
    expect(getByTestId('el')).toHaveStyleRule('height', '100%', {
      media: mediaQueries.phone,
    })
  })

  it('applies the "size" prop', () => {
    const {getByTestId} = renderMq(
      <Frame size={{phone: '100%'}} data-testid="el" />
    )
    expect(getByTestId('el')).toHaveStyleRule('width', '100%', {
      media: mediaQueries.phone,
    })
    expect(getByTestId('el')).toHaveStyleRule('height', '100%', {
      media: mediaQueries.phone,
    })
  })

  it('applies the "bg" prop', () => {
    const {getByTestId} = renderMq(
      <Frame bg={{phone: 'green'}} data-testid="el" />
    )
    expect(getByTestId('el')).toHaveStyleRule(
      'background',
      'var(--color-green)',
      {
        media: mediaQueries.phone,
      }
    )
  })

  it('applies the "pad" prop', () => {
    const {getByTestId} = renderMq(<Frame pad={{phone: 1}} data-testid="el" />)
    expect(getByTestId('el')).toHaveStyleRule('padding', 'var(--pad-1)', {
      media: mediaQueries.phone,
    })
  })

  it('applies the "pad" prop w/ array', () => {
    const {getByTestId} = renderMq(
      <Frame pad={{phone: [0, 'auto']}} data-testid="el" />
    )
    expect(getByTestId('el')).toHaveStyleRule(
      'padding',
      'var(--pad-0) var(--pad-auto)',
      {
        media: mediaQueries.phone,
      }
    )
  })

  it('applies the "radius" prop', () => {
    const {getByTestId} = renderMq(
      <Frame radius={{phone: 'sm'}} data-testid="el" />
    )
    expect(getByTestId('el')).toHaveStyleRule(
      'border-radius',
      'var(--radius-sm)',
      {
        media: mediaQueries.phone,
      }
    )
  })

  it('applies the "radius" prop w/ array', () => {
    const {getByTestId} = renderMq(
      <Frame radius={{phone: ['sm', 'md']}} data-testid="el" />
    )
    expect(getByTestId('el')).toHaveStyleRule(
      'border-radius',
      'var(--radius-sm) var(--radius-md)',
      {
        media: mediaQueries.phone,
      }
    )
  })

  it('applies the "elevation" prop', () => {
    const {getByTestId} = renderMq(
      <Frame elevation={{phone: 'low'}} data-testid="el" />
    )
    expect(getByTestId('el')).toHaveStyleRule(
      'box-shadow',
      'var(--elevation-low)',
      {
        media: mediaQueries.phone,
      }
    )
  })
})
