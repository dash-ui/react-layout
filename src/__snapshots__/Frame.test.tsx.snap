// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Frame> without media queries applies several props 1`] = `
<DocumentFragment>
  .-ui-0 {
  display: block;
}

.-ui-1 {
  position: absolute;
}

.-ui-2 {
  width: 200px;
  height: 200px;
}

.-ui-3 {
  padding: var(--pad-3);
}

.-ui-4 {
  background: var(--color-blue);
}

.-ui-5 {
  box-shadow: var(--elevation-high);
}

<div
    class="-ui-0 -ui-1 -ui-2 -ui-3 -ui-4 -ui-5"
    data-testid="el"
  />
</DocumentFragment>
`;

exports[`<Frame> without media queries should throw if media queries are used without <LayoutProvider> 1`] = `"@-ui/layout is attempting to use media queries without a <LayoutProvider> component. This is not allowed."`;
